lib/Importer/Caffe2ModelLoader.cpp

```diff
@@ -1078,14 +1078,29 @@ llvm::Error Caffe2ModelLoader::loadWeights(caffe2::NetDef &net) {
 Caffe2ModelLoader::Caffe2ModelLoader(const std::string &netDescFilename,
                                      const std::string &netWeightFilename,
                                      llvm::ArrayRef<const char *> names,
-                                     llvm::ArrayRef<TypeRef> types, Function &F)
-    : CommonOperatorLoader(names, types, F) {
-  // The caffe2 network descriptor that we are deserializing.
-  caffe2::NetDef networkDef = EXIT_ON_ERR(loadProtoFile(netDescFilename));
+                                     llvm::ArrayRef<TypeRef> types, Function &F,
+                                     llvm::Error *errPtr)
+    : CommonOperatorLoader(names, types, F, errPtr) {
+  // Lambda to setup the Caffe2ModelLoader and return any llvm::Errors that were
+  // raised.
+  auto setup = [&]() -> llvm::Error {
+    // The caffe2 network descriptor that we are deserializing.
+    caffe2::NetDef networkDef;
+    ASSIGN_VALUE_OR_RETURN_ERR(networkDef, loadProtoFile(netDescFilename));
+
+    // The caffe2 weights that we are deserializing.
+    caffe2::NetDef weightsDef;
+    ASSIGN_VALUE_OR_RETURN_ERR(weightsDef, loadProtoFile(netWeightFilename));
+
+    RETURN_IF_ERR(loadWeights(weightsDef));
+    RETURN_IF_ERR(loadNetwork(networkDef));
 
-  // The caffe2 weights that we are deserializing.
-  caffe2::NetDef weightsDef = EXIT_ON_ERR(loadProtoFile(netWeightFilename));
+    return llvm::Error::success();
+  };
 
-  TEMP_EXIT_ON_ERR(loadWeights(weightsDef));
-  TEMP_EXIT_ON_ERR(loadNetwork(networkDef));
+  if (errPtr) {
+    *errPtr = setup();
+  } else {
+    EXIT_ON_ERR(setup());
```

Comment:
Does this mean errors in importing onnx will terminate the whole program? 

https://github.com/pytorch/glow/blob/4014bde87ac1e45aec7a3a039ece1473dceff94a/include/glow/Support/Error.h#L48-L50

Reply:
Before having added this error handling stuff to onnx/c2 importing code any errors in the importer would have caused the whole program to terminate no matter what. Now most methods in the loader return an Error which can be handled and appropriately by the user. In this case, you can't return an Error from a constructor so you can either pass a pointer to an Error and any error that occurs during construction can be placed there or if the user chooses not to do that then importing will exit the program on error like before. This is strictly safer than before because now the user of the importer has the option to handle any Errors without killing the program whereas before the program would get killed on any error that occurred no matter what.

Reply:
Ok.  I got it. Thanks for explanation. Let's make sure the caller do supply this error pointer. :) Exiting the program on import exceptions is not desirable.